### PR TITLE
streamingccl: stream replication builtin "crdb_internal.start_replication_stream"

### PR DESCRIPTION
**Pull request by gh-casper**
1. Introduce ReplicationStreamManager through which stream replications builtins invoke streamingccl logic.
2. Support a new builtin "crdb_internal.start_replication_stream" that creates a producer job for specified tenant on the source cluster to track liveness of a stream replication.

Release note (sql change): introduce a builtin function "crdb_internal.start_replication_stream, a cluster setting "stream_replication.job_liveness_timeout" and a cluster setting "stream_replication.stream_liveness_track_frequency".

### CODE REVIEW

**Review by miretskiy — CHANGES_REQUESTED** [review]
This is very nice.  Few nits/questions.
@dt might have more comments.

> Reviewed 19 of 22 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-:-MnXaSvKEhJSqOrf2-9x:buzhfoy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper and @shermanCRL)

___
*[pkg/ccl/streamingccl/init.go, line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXDP3W36kyMLj6yHAa:-MnXDP3W36kyMLj6yHAb:bnwlpdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/init.go#L9)):*
> ```Go
> //     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
> 
> package streamingccl
> ```

Let's rename this file.  I don't think `init.go` is specific enough.
How about `replication_manager.go`?
___
*[pkg/ccl/streamingccl/init.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXQKyD3v8Aq1EkDFFx:-MnXQKyD3v8Aq1EkDFFy:b3tso6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/init.go#L24)):*
> ```Go
> // CompleteStreamIngestionHook hooks a CompleteStreamIngestion implementation
> // inside streamingccl package.
> var CompleteStreamIngestionHook func(evalCtx *tree.EvalContext, txn *kv.Txn, jobID int, cutoverTimestamp hlc.Timestamp) error
> ```

Since you invoke methods (CompleteStreamIngestion, as well as InitStreamHook) on the replication stream manager, I wonder
if we should just drop these hooks?

Simply keep GetReplicationStreamManagerHook  as the only hook function, and simply implement InitStream/CompleteStreamIngestion on the manager
instance?

I also don't know if we should be bringing in CompleteStreamIngestionHook under this manager.
I wonder if we should separate APIs for producer cluster from those for the consumer cluster.
@dt wdyt?
___
*[pkg/ccl/streamingccl/init.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXDanL89oKMghsVYG_:-MnXDanL89oKMghsVYGa:b27rox1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/init.go#L49)):*
> ```Go
> func init() {
> 	streaming.GetReplicationStreamManagerHook = func() (streaming.ReplicationStreamManager, error) {
> 		return replicationStreamManagerImpl{}, nil
> ```

Let's return `&replicationStreamManager` instead.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXRdzCDI5EQIMDs-Gt:-MnXRdzD7N17B--VDBe-:b-71v0p0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/streamproducer/producer_job.go#L94)):*
> ```Go
> // 1. Tracks the liveness of the replication stream consumption
> // 2. TODO(casper): Updates the protected timestamp for spans being replicated
> func doInitStream(evalCtx *tree.EvalContext, txn *kv.Txn, tenantID uint64) (jobspb.JobID, error) {
> ```

do you need to rebase on master? Looks like this code was already committed?
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 96 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXXqOo19q3hT8wjINZ:-MnXXqOo19q3hT8wjIN_:b-f7rdds) ([raw file](https://github.com/cockroachdb/cockroach/blob/23700dfbd0baa3d4e86cb484f3552daf3c866a3d/pkg/ccl/streamingccl/streamproducer/producer_job.go#L96)):*
> ```Go
> func doInitStream(evalCtx *tree.EvalContext, txn *kv.Txn, tenantID uint64) (jobspb.JobID, error) {
> 	username := evalCtx.Username()
> 	hasAdminRole, err := evalCtx.Planner.UserHasAdminRole(evalCtx.Ctx(), username)
> ```

I don't think this is necessary (and neither is exposing username method directly on evalCtx.
See `evalCtx.SessionData().User()` and `evalCtx.SessionData().HasAdminRole`
___
*[pkg/ccl/streamingccl/streamproducer/replication_stream_test.go, line 205 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXRr4OCMnsuYd2yMa7:-MnXRr4OCMnsuYd2yMa8:b-ot459h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L205)):*
> ```Go
> 
> 		time.Sleep(500 * time.Millisecond) // Wait for the job status transitioning
> 		h.SysDB.SucceedsSoonDuration = 2 * time.Second
> ```

Please: no sleeps in test.
___
*[pkg/ccl/streamingccl/streamproducer/replication_stream_test.go, line 216 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXYqj1Bl11Y-NuQBUW:-MnXYqj1Bl11Y-NuQBUX:b67i11p) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L216)):*
> ```Go
> 
> 		time.Sleep(500 * time.Millisecond) // Wait for the job status transitioning
> 		testDuration, now := 2*time.Second, timeutil.Now()
> ```

no sleep in tests please.
___
*[pkg/sql/conn_executor.go, line 2466 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXaDRMEkXUiYBSSI09:-MnXaDRMEkXUiYBSSI0A:bfkh6vo) ([raw file](https://github.com/cockroachdb/cockroach/blob/23700dfbd0baa3d4e86cb484f3552daf3c866a3d/pkg/sql/conn_executor.go#L2466)):*
> ```Go
> 			ReCache:                   ex.server.reCache,
> 			InternalExecutor:          &ie,
> 			Username:                  p.User,
> ```

Probably not needed.
___
*[pkg/sql/sem/builtins/builtins.go, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXBCo_Adx1EewIr8nD:-MnXBCo_Adx1EewIr8nE:b-qztsio) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/sql/sem/builtins/builtins.go#L128)):*
> ```Go
> 	categorySystemRepair        = "System repair"
> 	categoryStreamIngestion     = "Stream Ingestion"
> 	categoryStreamReplication   = "Stream Replication"
> ```

do we need to introduce new category? Why not use stream ingestion?
___
*[pkg/sql/sem/builtins/replication_builtins.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXBaRX-Q19wtXdkP3T:-MnXBaRX-Q19wtXdkP3U:b-dm754e) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/sql/sem/builtins/replication_builtins.go#L69)):*
> ```Go
> 	),
> 
> 	"crdb_internal.init_stream": makeBuiltin(
> ```

nit: not a fan of init_stream name.  Can we re name e.g. `start_replication_stream`
___
*[pkg/sql/sem/builtins/replication_builtins.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXYejIEQ7oKfacsJJ_:-MnXYejIEQ7oKfacsJJa:bcqos1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/sql/sem/builtins/replication_builtins.go#L82)):*
> ```Go
> 				mgr, err := streaming.GetReplicationStreamManager()
> 				if err != nil {
> 					return tree.NewDInt(tree.DInt(0)), err
> ```

can you return `nil, err`?
___
*[pkg/sql/sem/builtins/replication_builtins.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXBv-R7nk7OJtZ35_k:-MnXBv-S3KoNDADSDbh_:b-n80xgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/sql/sem/builtins/replication_builtins.go#L85)):*
> ```Go
> 				}
> 
> 				tenantID := int(*args[0].(*tree.DInt))
> ```

Let's use
```
tenanID := int(tree.MustBeDint(args[0]))
```
___
*[pkg/sql/sem/builtins/replication_builtins.go, line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXCFwZCNN2Hc-j_LzJ:-MnXCFwZCNN2Hc-j_LzK:b-94qh6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/sql/sem/builtins/replication_builtins.go#L92)):*
> ```Go
> 				"on the producer side. The job will periodically check liveness of the stream replication " +
> 				"and will kill the job if it has been inactive for a duration of time specified by the " +
> 				"cluster setting 'stream_replication.job_liveness_timeout'.",
> ```

nit: I think we should describe what the function does, and not necessarily describe implementation details (such as job liveness).
How about:

```
This function can be used on the producer side to start a replication stream for the specified tenant.
The returned stream ID uniquely identifies create stream.  The caller must periodically invoke
crdb_internal.heartbeat_stream() function to notify that the replication is still ongoing.
```
___
*[pkg/streaming/utils.go, line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXB-TIEWJPTso-ur8c:-MnXB-TIEWJPTso-ur8d:bneghkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/streaming/utils.go#L26)):*
> ```Go
> // ReplicationStreamManager represents a collection of APIs that streaming replication supports.
> type ReplicationStreamManager interface {
> ```

nit: extra \n?
___
*[pkg/streaming/utils.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnX4N8jDhRqZw8InhaR:-MnX4N8jDhRqZw8InhaS:b-e9v6la) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/streaming/utils.go#L28)):*
> ```Go
> 
> 	// CompleteStreamIngestion signals a running stream ingestion job to complete.
> 	CompleteStreamIngestion(evalCtx *tree.EvalContext, txn *kv.Txn, jobID int, cutoverTimestamp hlc.Timestamp) error
> ```

Let's not use jobID -- it's an internal implementation detail.
Let's introduce 
```
type StreamID int64
```

and use StreamID throughout.
___
*[pkg/streaming/utils.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnX4InDAXF0SWMmJvnJ:-MnX4InEBkYc8OpJ6x9s:bah9kzc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/streaming/utils.go#L31)):*
> ```Go
> 
> 	// InitStream starts a stream replication job for the specified tenant on the producer side.
> 	InitStream(evalCtx *tree.EvalContext, txn *kv.Txn, tenantID uint64) (jobspb.JobID, error)
> ```

nit: Let's rename to StartStream()?
___
*[pkg/streaming/utils.go, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXAefh1CBQQCzFl_Ya:-MnXAefh1CBQQCzFl_Yb:b3a0wao) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/streaming/utils.go#L37)):*
> ```Go
> func GetReplicationStreamManager() (ReplicationStreamManager, error) {
> 	if GetReplicationStreamManagerHook == nil {
> 		return nil, errors.New("completing a stream replication job requires a CCL binary")
> ```

nit: error message needs to be updated.
(`replication streaming requires CCL binary`)?

**Review by gh-casper — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72330)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, and @shermanCRL)

___
*[pkg/ccl/streamingccl/init.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXQKyD3v8Aq1EkDFFx:-MnXcv8m8DwePotBQh2q:b8dprmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/init.go#L24)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Since you invoke methods (CompleteStreamIngestion, as well as InitStreamHook) on the replication stream manager, I wonder
if we should just drop these hooks?

Simply keep GetReplicationStreamManagerHook  as the only hook function, and simply implement InitStream/CompleteStreamIngestion on the manager
instance?

I also don't know if we should be bringing in CompleteStreamIngestionHook under this manager.
I wonder if we should separate APIs for producer cluster from those for the consumer cluster.
@dt wdyt?
</blockquote></details>

I want to mention my thoughts on this before next revision:

This manager implementation(replicationStreamManagerImpl) is under streamingccl, but doCompleteIngestion is under streamingutils and doInitStream is under streamproducer.  I personally also try my best to make code as clean as possible, but here it looks like I can only use composition pattern (installing hooks inside replicationStreamManagerImpl) instead of implementing the ReplicationStreamManager directly. 
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXRdzCDI5EQIMDs-Gt:-MnXc3dp4RVA8_DMuKeQ:balb8dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/streamproducer/producer_job.go#L94)):*
> oks like this
No, this is based on master. I delete this function in the previous PR.

**Review by miretskiy — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72330)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, and @shermanCRL)

___
*[pkg/ccl/streamingccl/init.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXQKyD3v8Aq1EkDFFx:-MnXpqn-2tYR6D-FtqHK:b5k6m5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/init.go#L24)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

I want to mention my thoughts on this before next revision:

This manager implementation(replicationStreamManagerImpl) is under streamingccl, but doCompleteIngestion is under streamingutils and doInitStream is under streamproducer.  I personally also try my best to make code as clean as possible, but here it looks like I can only use composition pattern (installing hooks inside replicationStreamManagerImpl) instead of implementing the ReplicationStreamManager directly. 
</blockquote></details>

Sure... But can we perhaps avoid these hooks? Perhaps this particular code should act as a bridge between sql tree and the rest of streaming ccl code.
So, we have manager hook install so that crdb_internal functions can access that; but beyond that, why can't we directly
call functions in question?  e.g Instead of having createreplicationhook you just call e.g. streamingccl.CreateReplication() method?

**Review by gh-casper — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72330)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @shermanCRL)

___
*[pkg/ccl/streamingccl/init.go, line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXDP3W36kyMLj6yHAa:-MnaRE8WBtsjyKEg30Yx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/init.go#L9)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Let's rename this file.  I don't think `init.go` is specific enough.
How about `replication_manager.go`?
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/init.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXQKyD3v8Aq1EkDFFx:-MnaRHnE5miYupt5bz7z:b-rhmlbk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/init.go#L24)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Sure... But can we perhaps avoid these hooks? Perhaps this particular code should act as a bridge between sql tree and the rest of streaming ccl code.
So, we have manager hook install so that crdb_internal functions can access that; but beyond that, why can't we directly
call functions in question?  e.g Instead of having createreplicationhook you just call e.g. streamingccl.CreateReplication() method?
</blockquote></details>

We can't call streamproducer.StartReplicationStream under streamingccl because it will introduce cycle imports.
___
*[pkg/ccl/streamingccl/init.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXDanL89oKMghsVYG_:-MnaRYGj5lRA36Y_7nPK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/init.go#L49)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Let's return `&replicationStreamManager` instead.
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 96 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXXqOo19q3hT8wjINZ:-MnaSg7m2ZFc_2_DzyOd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/23700dfbd0baa3d4e86cb484f3552daf3c866a3d/pkg/ccl/streamingccl/streamproducer/producer_job.go#L96)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I don't think this is necessary (and neither is exposing username method directly on evalCtx.
See `evalCtx.SessionData().User()` and `evalCtx.SessionData().HasAdminRole`

</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/replication_stream_test.go, line 205 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXRr4OCMnsuYd2yMa7:-Mnayqh97m-ZiFjSfBZQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L205)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Please: no sleeps in test.
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/replication_stream_test.go, line 216 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXYqj1Bl11Y-NuQBUW:-MnayqzcEkou3BWgStzh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L216)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

no sleep in tests please.
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor.go, line 2466 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXaDRMEkXUiYBSSI09:-MnaSt1b94TjFA57EJgp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/23700dfbd0baa3d4e86cb484f3552daf3c866a3d/pkg/sql/conn_executor.go#L2466)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Probably not needed.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/builtins.go, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXBCo_Adx1EewIr8nD:-MnaTNSN0hmLwDSPB4mr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/sql/sem/builtins/builtins.go#L128)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

do we need to introduce new category? Why not use stream ingestion?
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/replication_builtins.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXBaRX-Q19wtXdkP3T:-MnaTi6MCOGdt9aUeMKV:b-8atv9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/sql/sem/builtins/replication_builtins.go#L69)):*
> start_replication_stream
Done.
___
*[pkg/sql/sem/builtins/replication_builtins.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXYejIEQ7oKfacsJJ_:-MnaTvpd90BZq-GPKLTQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/sql/sem/builtins/replication_builtins.go#L82)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

can you return `nil, err`?
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/replication_builtins.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXBv-R7nk7OJtZ35_k:-MnaUCdl5AlOFmSlUmf-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/sql/sem/builtins/replication_builtins.go#L85)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Let's use
```
tenanID := int(tree.MustBeDint(args[0]))
```
</blockquote></details>

Done.
___
*[pkg/streaming/utils.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnX4N8jDhRqZw8InhaR:-Mnayoo3BZyk34gp9O45:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/streaming/utils.go#L28)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Let's not use jobID -- it's an internal implementation detail.
Let's introduce 
```
type StreamID int64
```

and use StreamID throughout.
</blockquote></details>

Done.

File: pkg/ccl/streamingccl/streamproducer/replication_stream_test.go [hunk 1]
```diff
@@ -198,21 +198,21 @@ func TestReplicationStreamInitialization(t *testing.T) {
 
 	t.Run("failed-after-timeout", func(t *testing.T) {
 		h.SysDB.Exec(t, "SET CLUSTER SETTING stream_replication.job_liveness_timeout = '1s'")
-		rows := h.SysDB.QueryStr(t, "SELECT crdb_internal.init_stream($1)", h.Tenant.ID.ToUint64())
+		rows := h.SysDB.QueryStr(t, "SELECT crdb_internal.start_replication_stream($1)", h.Tenant.ID.ToUint64())
 		jobID := rows[0][0]
 
-		time.Sleep(500 * time.Millisecond) // Wait for the job status transitioning
+		//time.Sleep(500 * time.Millisecond) // Wait for the job status transitioning
```
Comment (miretskiy) [inline pkg/ccl/streamingccl/streamproducer/replication_stream_test.go:204]: Drop this comment?
Is succeedssoonduration of 2 seconds enough? Even when running under `-race`?

File: pkg/ccl/streamingccl/replication_manager.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/cockroachdb/errors"
+)
+
+type replicationStreamManagerImpl struct{}
+
+// CompleteStreamIngestionHook hooks a CompleteStreamIngestion implementation
+// inside streamingccl package.
+var CompleteStreamIngestionHook func(evalCtx *tree.EvalContext, txn *kv.Txn, jobID int, cutoverTimestamp hlc.Timestamp) error
+
+func (r replicationStreamManagerImpl) CompleteStreamIngestion(
```
Comment (miretskiy) [inline pkg/ccl/streamingccl/replication_manager.go:25]: nit:
lets add method comments here and start method:

// CompleteStreamIngestion implements ReplicationStreamManager interface.

**Review by gh-casper — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72330)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @shermanCRL)

___
*[pkg/ccl/streamingccl/streamproducer/replication_stream_test.go, line 204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnWknSZEriJz6N4W3Af-r3-204:-Mng0nR34jJbPmTj69RA:b3j1l17) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e92f6158d53b345f95d2beb02126f1fe981f859/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L204)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Drop this comment?
Is succeedssoonduration of 2 seconds enough? Even when running under `-race`?
</blockquote></details>

Changed it to 5 seconds.

File: pkg/sql/sem/builtins/replication_builtins.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+			ReturnType: tree.FixedReturnType(types.Int),
+			Fn: func(evalCtx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+				mgr, err := streaming.GetReplicationStreamManager()
+				if err != nil {
+					return nil, err
+				}
+
+				tenantID := int(tree.MustBeDInt(args[0]))
+				jobID, err := mgr.StartReplicationStream(evalCtx, evalCtx.Txn, uint64(tenantID))
+				return tree.NewDInt(tree.DInt(jobID)), err
```
Comment (miretskiy) [inline pkg/sql/sem/builtins/replication_builtins.go:87]: Let's not save typing; let's add
`if err != nil { return nil, err }`

File: pkg/sql/sem/builtins/replication_builtins.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+				mgr, err := streaming.GetReplicationStreamManager()
+				if err != nil {
+					return nil, err
+				}
+
+				jobID := int(*args[0].(*tree.DInt))
+				cutoverTime := args[1].(*tree.DTimestampTZ).Time
+				cutoverTimestamp := hlc.Timestamp{WallTime: cutoverTime.UnixNano()}
+				err = mgr.CompleteStreamIngestion(evalCtx, evalCtx.Txn, jobID, cutoverTimestamp)
+				return tree.NewDInt(tree.DInt(jobID)), err
```
Comment (miretskiy) [inline pkg/sql/sem/builtins/replication_builtins.go:55]: Let's add
if err != nil {return nil, err}

File: pkg/ccl/streamingccl/streamproducer/replication_stream_test.go [hunk 5]
```diff
@@ -175,3 +179,42 @@ INSERT INTO d.t2 VALUES (2);
...
+	})
+
+	t.Run("continuously-running-within-timeout", func(t *testing.T) {
+		h.SysDB.Exec(t, "SET CLUSTER SETTING stream_replication.job_liveness_timeout = '10s'")
+		rows := h.SysDB.QueryStr(t, "SELECT crdb_internal.start_replication_stream($1)", h.Tenant.ID.ToUint64())
+		jobID := rows[0][0]
+
+		testDuration, now := 2*time.Second, timeutil.Now()
+		for start, end := now, now.Add(testDuration); start.Before(end); start = start.Add(300 * time.Millisecond) {
+			h.SysDB.CheckQueryResults(t, fmt.Sprintf("SELECT status FROM system.jobs WHERE id = %s", jobID),
```
[the remark this answers is not in the document]
Comment (gh-casper) [reply pkg/ccl/streamingccl/streamproducer/replication_stream_test.go:217]: What I'm trying to do here is to make sure the job is still "running" for several seconds. I just added a comment above. Before this, I also wait until it's running before looping for 3 seconds.

**Review by miretskiy — APPROVED** [review]
> Reviewed 12 of 14 files at r3, 3 of 3 files at r4, 3 of 4 files at r5, 6 of 6 files at r6, 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72330)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, and @shermanCRL)

___
*[pkg/ccl/streamingccl/init.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnXQKyD3v8Aq1EkDFFx:-Mngd3tN6yMC0qpgf0Na:b-9uj6me) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a8f9fb6423a309f832729d8c9f97b3bcc3373/pkg/ccl/streamingccl/init.go#L24)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

We can't call streamproducer.StartReplicationStream under streamingccl because it will introduce cycle imports.
</blockquote></details>

Probably need to further refactor/move the code around.  I honestly don't like it that much to have these hooks for each method we need to add.  I think it's okay as a follow on though.
___
*[pkg/ccl/streamingccl/streamproducer/replication_stream_test.go, line 204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72330#-MnWknSZEriJz6N4W3Af-r3-204:-MngcarP30up_0GW7xbt:bzbifti) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e92f6158d53b345f95d2beb02126f1fe981f859/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L204)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

Changed it to 5 seconds.
</blockquote></details>

I don't understand why we need it to be overriden at all.
Why not use SucceedsSoon instead?  When all is good -- that should definitely succeed
under 45 seconds.  Do you really want this test to flake if it's running under race?
Of course, if the test is stuck and you're developing, just changing the test to lower the timeout
makes sense; but I'm not sure you should keep this lower setting in general.
Ditto below.

**Pull request merged**